src/vpn-api/worker.ts

```diff
@@ -143,13 +162,77 @@ const worker = (instanceId: number) => {
 			),
 		)
 		.tap(() => openvpn.start())
+		.delay(1000)
+		.tap(() => {
+			return vpn.connect({ port: mgtPort, shellPrompt: '' }).then(() => {
+				vpn.connection.shellPrompt = '';
+				// monitor new client connections and map cid to uuid
+				vpn.on('log', data => {
+					if (data.includes('CLIENT:ESTABLISHED,')) {
+						const clientId = data.split(',')[1].trim();
+						const idMapper = (logData: string) => {
+							if (logData.includes('CLIENT:ENV,common_name=')) {
+								const now = new Date().getTime();
+								const uuid = logData.split('=')[1].trim();
+								// expire any previous mappings for this device
+								_.forOwn(clients, (client, key) => {
+									if (client.uuid === uuid) {
+										delete clients[key];
+									}
+								});
+								// register current mapping
+								clients[clientId] = {
+									uuid,
+									ts: now,
+									heartbeat: now,
+									tx: 0,
+									rx: 0,
+								};
+								logger.info(
+									`Parsed connect event for client_id=${clientId} uuid=${uuid}`,
+								);
+								vpn.removeListener('log', idMapper);
+							}
+						};
+						vpn.on('log', idMapper);
+					}
+				});
+				// process bytecount events to track realtime data usage
+				vpn.on('data', data => {
+					if (data.bytecount == null) {
+						return;
+					}
+					if (!data.bytecount[0].includes('CLI:')) {
```

Comment:
Is the `CLI:` guaranteed to be at the start?  I'm wondering because you split on `:` and take the second element below so I'm expecting it to be to extract this.  If so it might be better to do the split and do an `===` on the first element

Reply:
yes, the actual format is `>BYTECOUNT_CLI:` but the `>BYTECOUNT` part is trimmed from the library (which is admittedly a bit hacky, the library doesn't seem to be expecting CLI bytecount messages, just `>BYTECOUNT` messages which are global counters, but it works -- I guess ideally we would extend the openvpn-telnet library to know about bytecount_cli messages).

Reply:
I've PR'd `telnet-openvpn`, if it doesn't get a response soon-ish I'll just fork the repo and publish under `@balena`

Reply:
Nice, he’s merged it and released a new version. I’ll get this branch updated.